### ISSUE TITLE
Missed optimization on only used once allocation

### ISSUE THREAD

**Issue by lzutao** [rust-lang/rust#67199]
godbolt link: https://rust.godbolt.org/z/d_VvSo
Here are two snippets that I expect that they should have the same assembly:
```rust
pub fn foo(s: &str) -> bool {
    s != "FOOFOOLONGLONG"
}
```

```rust
pub fn foo(s: &str) -> bool {
     s.to_string() != "FOOFOOLONGLONG"
}
```

**Comment by lzutao** [comment]
The problem with this is that the second snippet is used rather often.
For example:
```rust
pub fn foo() -> bool {
    std::env::var("FOO").unwrap_or("0".to_string()) != "BAR"
}
```

**Comment by Ixrec** [comment]
FWIW, there's [already a clippy lint](https://rust-lang.github.io/rust-clippy/master/#or_fun_call) to tell you to replace things like `unwrap_or(...)` with `unwrap_or_else(|| { ... })`. for precisely this reason. This is also part of the reason why so many Option/Result methods have a _then or _else variation that takes a closure instead of a value.

**Comment by lzutao** [comment]
Yes, the version with unwrap_or_else has better codegen than one with unwrap_or, 
but they still has worse codegen than the one doesn't allocate.
